// zotero-plugin-toolkit/dist/helpers/progressWindow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressWindowHelper = void 0;
const basic_1 = require("../basic");
/**
 * ProgressWindow helper.
 * @example
 * Show a popup with success icon
 * ```ts
 * const tool = new ZoteroTool();
 * tool.createProgressWindow("Addon").createLine({
 *   type: "success",
 *   text: "Finish"
 *   progress: 100,
 * }).show();
 * ```
 * @example
 * Show a popup and change line content
 * ```ts
 * const compat = new ZoteroCompat();
 * const tool = new ZoteroTool();
 * const popupWin = tool.createProgressWindow("Addon").createLine({
 *   text: "Loading"
 *   progress: 50,
 * }).show(-1);
 * // Do operations
 * compat.getGlobal("setTimeout")(()=>{
 *   popupWin.changeLine({
 *     text: "Finish",
 *     progress: 100,
 *   });
 * }, 3000);
 * ```
 */
class ProgressWindowHelper extends Zotero.ProgressWindow {
    /**
     *
     * @param header window header
     * @param options
     */
    constructor(header, options = {
        closeOnClick: true,
        closeTime: 5000,
    }) {
        super(options);
        this.lines = [];
        this.closeTime = options.closeTime || 5000;
        this.changeHeadline(header);
        this.originalShow = this
            .show;
        this.show = this.showWithTimer;
        if (options.closeOtherProgressWindows) {
            basic_1.BasicTool.getZotero().ProgressWindowSet.closeAll();
        }
    }
    /**
     * Create a new line
     * @param options
     */
    createLine(options) {
        const icon = this.getIcon(options.type, options.icon);
        const line = new this.ItemProgress(icon || "", options.text || "");
        if (typeof options.progress === "number") {
            line.setProgress(options.progress);
        }
        this.lines.push(line);
        this.updateIcons();
        return this;
    }
    /**
     * Change the line content
     * @param options
     */
    changeLine(options) {
        var _a;
        if (((_a = this.lines) === null || _a === void 0 ? void 0 : _a.length) === 0) {
            return this;
        }
        const idx = typeof options.idx !== "undefined" &&
            options.idx >= 0 &&
            options.idx < this.lines.length
            ? options.idx
            : 0;
        const icon = this.getIcon(options.type, options.icon);
        if (icon) {
            // @ts-ignore
            this.lines[idx].setItemTypeAndIcon(icon);
        }
        options.text && this.lines[idx].setText(options.text);
        typeof options.progress === "number" &&
            this.lines[idx].setProgress(options.progress);
        this.updateIcons();
        return this;
    }
    showWithTimer(closeTime = undefined) {
        this.originalShow();
        typeof closeTime !== "undefined" && (this.closeTime = closeTime);
        if (this.closeTime && this.closeTime > 0) {
            this.startCloseTimer(this.closeTime);
        }
        setTimeout(this.updateIcons.bind(this), 50);
        return this;
    }
    /**
     * Set custom icon uri for progress window
     * @param key
     * @param uri
     */
    static setIconURI(key, uri) {
        icons[key] = uri;
    }
    getIcon(type, defaultIcon) {
        return type && type in icons ? icons[type] : defaultIcon;
    }
    updateIcons() {
        try {
            this.lines.forEach((line) => {
                const box = line._image;
                const icon = box.dataset.itemType;
                if (icon &&
                    icon.startsWith("chrome://") &&
                    !box.style.backgroundImage.includes("progress_arcs")) {
                    box.style.backgroundImage = `url(${box.dataset.itemType})`;
                }
            });
        }
        catch (e) {
            // Ignore
        }
    }
}
exports.ProgressWindowHelper = ProgressWindowHelper;
/**
 * Icon dict. Add your custom icons here.
 * @default
 * ```ts
 * {
 *   success: "chrome://zotero/skin/tick.png",
 *   fail: "chrome://zotero/skin/cross.png",
 * };
 * ```
 */
const icons = {
    success: "chrome://zotero/skin/tick.png",
    fail: "chrome://zotero/skin/cross.png",
};
