// zotero-plugin-toolkit/dist/managers/fieldHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldHookManager = void 0;
const patch_1 = require("../helpers/patch");
const basic_1 = require("../basic");
/**
 * Item field hooks manager.
 */
class FieldHookManager extends basic_1.ManagerTool {
    constructor(base) {
        super(base);
        this.data = {
            getField: {},
            setField: {},
            isFieldOfBase: {},
        };
        this.patchHelpers = {
            getField: new patch_1.PatchHelper(),
            setField: new patch_1.PatchHelper(),
            isFieldOfBase: new patch_1.PatchHelper(),
        };
        const _thisHelper = this;
        for (const type of Object.keys(this.patchHelpers)) {
            const helper = this.patchHelpers[type];
            helper.setData({
                target: this.getGlobal("Zotero").Item.prototype,
                funcSign: type,
                patcher: (original) => function (field, ...args) {
                    // @ts-ignore
                    const originalThis = this;
                    const handler = _thisHelper.data[type][field];
                    if (typeof handler === "function") {
                        try {
                            // @ts-ignore
                            return handler(field, args[0], args[1], originalThis, original);
                        }
                        catch (e) {
                            return field + String(e);
                        }
                    }
                    // @ts-ignore
                    return original.apply(originalThis, [field, ...args]);
                },
                enabled: true,
            });
        }
    }
    register(type, field, hook) {
        this.data[type][field] = hook;
    }
    unregister(type, field) {
        delete this.data[type][field];
    }
    unregisterAll() {
        this.data.getField = {};
        this.data.setField = {};
        this.data.isFieldOfBase = {};
        this.patchHelpers.getField.disable();
        this.patchHelpers.setField.disable();
        this.patchHelpers.isFieldOfBase.disable();
    }
}
exports.FieldHookManager = FieldHookManager;
