// zotero-plugin-toolkit/dist/managers/menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuManager = void 0;
const ui_1 = require("../tools/ui");
const basic_1 = require("../basic");
/**
 * Register \<menuitem\>, \<menupopup\>, or \<menuseperator\> to Zotero right-click/window menus.
 */
class MenuManager extends basic_1.ManagerTool {
    constructor(base) {
        super(base);
        this.ui = new ui_1.UITool(this);
    }
    /**
     * Insert an menu item/menu(with popup)/menuseprator into a menupopup
     * @remarks
     * options:
     * ```ts
     * export interface MenuitemOptions {
     *   tag: "menuitem" | "menu" | "menuseparator";
     *   id?: string;
     *   label?: string;
     *   // data url (chrome://xxx.png) or base64 url (data:image/png;base64,xxx)
     *   icon?: string;
     *   class?: string;
     *   styles?: { [key: string]: string };
     *   hidden?: boolean;
     *   disabled?: boolean;
     *   oncommand?: string;
     *   commandListener?: EventListenerOrEventListenerObject;
     *   // Attributes below are used when type === "menu"
     *   popupId?: string;
     *   onpopupshowing?: string;
     *   subElementOptions?: Array<MenuitemOptions>;
     * }
     * ```
     * @param menuPopup
     * @param options
     * @param insertPosition
     * @param anchorElement The menuitem will be put before/after `anchorElement`. If not set, put at start/end of the menupopup.
     * @example
     * Insert menuitem with icon into item menupopup
     * ```ts
     * // base64 or chrome:// url
     * const menuIcon = "chrome://addontemplate/content/icons/favicon@0.5x.png";
     * ztoolkit.Menu.register("item", {
     *   tag: "menuitem",
     *   id: "zotero-itemmenu-addontemplate-test",
     *   label: "Addon Template: Menuitem",
     *   oncommand: "alert('Hello World! Default Menuitem.')",
     *   icon: menuIcon,
     * });
     * ```
     * @example
     * Insert menu into file menupopup
     * ```ts
     * ztoolkit.Menu.register(
     *   "menuFile",
     *   {
     *     tag: "menu",
     *     label: "Addon Template: Menupopup",
     *     subElementOptions: [
     *       {
     *         tag: "menuitem",
     *         label: "Addon Template",
     *         oncommand: "alert('Hello World! Sub Menuitem.')",
     *       },
     *     ],
     *   },
     *   "before",
     *   Zotero.getMainWindow().document.querySelector(
     *     "#zotero-itemmenu-addontemplate-test"
     *   )
     * );
     * ```
     */
    register(menuPopup, options, insertPosition = "after", anchorElement) {
        let popup;
        if (typeof menuPopup === "string") {
            popup = this.getGlobal("document").querySelector(MenuSelector[menuPopup]);
        }
        else {
            popup = menuPopup;
        }
        if (!popup) {
            return false;
        }
        const doc = popup.ownerDocument;
        const genMenuElement = (menuitemOption) => {
            var _a, _b;
            const elementOption = {
                tag: menuitemOption.tag,
                id: menuitemOption.id,
                namespace: "xul",
                attributes: {
                    label: menuitemOption.label || "",
                    hidden: Boolean(menuitemOption.hidden),
                    disaled: Boolean(menuitemOption.disabled),
                    class: menuitemOption.class || "",
                    oncommand: menuitemOption.oncommand || "",
                },
                classList: menuitemOption.classList,
                styles: menuitemOption.styles || {},
                listeners: [],
                children: [],
            };
            if (menuitemOption.icon) {
                if (!this.getGlobal("Zotero").isMac) {
                    if (menuitemOption.tag === "menu") {
                        elementOption.attributes["class"] += " menu-iconic";
                    }
                    else {
                        elementOption.attributes["class"] += " menuitem-iconic";
                    }
                }
                elementOption.styles["list-style-image"] = `url(${menuitemOption.icon})`;
            }
            if (menuitemOption.commandListener) {
                (_a = elementOption.listeners) === null || _a === void 0 ? void 0 : _a.push({
                    type: "command",
                    listener: menuitemOption.commandListener,
                });
            }
            const menuItem = this.ui.createElement(doc, menuitemOption.tag, elementOption);
            if (menuitemOption.getVisibility) {
                popup.addEventListener("popupshowing", (ev) => {
                    const showing = menuitemOption.getVisibility(menuItem, ev);
                    if (showing) {
                        menuItem.removeAttribute("hidden");
                    }
                    else {
                        menuItem.setAttribute("hidden", "true");
                    }
                });
            }
            if (menuitemOption.tag === "menu") {
                const subPopup = this.ui.createElement(doc, "menupopup", {
                    id: menuitemOption.popupId,
                    attributes: { onpopupshowing: menuitemOption.onpopupshowing || "" },
                });
                (_b = menuitemOption.children) === null || _b === void 0 ? void 0 : _b.forEach((childOption) => {
                    subPopup.append(genMenuElement(childOption));
                });
                menuItem.append(subPopup);
            }
            return menuItem;
        };
        const topMenuItem = genMenuElement(options);
        if (!anchorElement) {
            anchorElement = (insertPosition === "after"
                ? popup.lastElementChild
                : popup.firstElementChild);
        }
        anchorElement[insertPosition](topMenuItem);
    }
    unregister(menuId) {
        var _a;
        (_a = this.getGlobal("document").querySelector(`#${menuId}`)) === null || _a === void 0 ? void 0 : _a.remove();
    }
    unregisterAll() {
        this.ui.unregisterAll();
    }
}
exports.MenuManager = MenuManager;
var MenuSelector;
(function (MenuSelector) {
    MenuSelector["menuFile"] = "#menu_FilePopup";
    MenuSelector["menuEdit"] = "#menu_EditPopup";
    MenuSelector["menuView"] = "#menu_viewPopup";
    MenuSelector["menuGo"] = "#menu_goPopup";
    MenuSelector["menuTools"] = "#menu_ToolsPopup";
    MenuSelector["menuHelp"] = "#menu_HelpPopup";
    MenuSelector["collection"] = "#zotero-collectionmenu";
    MenuSelector["item"] = "#zotero-itemmenu";
})(MenuSelector || (MenuSelector = {}));
