// zotero-plugin-toolkit/dist/utils/debugBridge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugBridge = void 0;
const basic_1 = require("../basic");
const toolkitGlobal_1 = __importDefault(require("../managers/toolkitGlobal"));
/**
 * Debug bridge.
 *
 * @remarks
 * Global variables: Zotero, window(main window).
 *
 * @example
 * Run script directly. The `run` is URIencoded script.
 *
 * `zotero://ztoolkit-debug/?run=Zotero.getMainWindow().alert(%22HelloWorld!%22)&app=developer`
 * @example
 * Run script from file. The `file` is URIencoded path to js file starts with `file:///`
 *
 * `zotero://ztoolkit-debug/?file=file%3A%2F%2F%2FC%3A%2FUsers%2Fw_xia%2FDesktop%2Frun.js&app=developer`
 */
class DebugBridge {
    get version() {
        return DebugBridge.version;
    }
    get disableDebugBridgePassword() {
        return this._disableDebugBridgePassword;
    }
    set disableDebugBridgePassword(value) {
        this._disableDebugBridgePassword = value;
    }
    get password() {
        return basic_1.BasicTool.getZotero().Prefs.get(DebugBridge.passwordPref, true);
    }
    set password(v) {
        basic_1.BasicTool.getZotero().Prefs.set(DebugBridge.passwordPref, v, true);
    }
    constructor() {
        this._disableDebugBridgePassword = false;
        this.initializeDebugBridge();
    }
    static setModule(instance) {
        var _a;
        if (!((_a = instance.debugBridge) === null || _a === void 0 ? void 0 : _a.version) ||
            instance.debugBridge.version < DebugBridge.version) {
            instance.debugBridge = new DebugBridge();
        }
    }
    initializeDebugBridge() {
        const debugBridgeExtension = {
            noContent: true,
            doAction: async (uri) => {
                var _a;
                const Zotero = basic_1.BasicTool.getZotero();
                const window = Zotero.getMainWindow();
                const uriString = uri.spec.split("//").pop();
                if (!uriString) {
                    return;
                }
                const params = {};
                (_a = uriString
                    .split("?")
                    .pop()) === null || _a === void 0 ? void 0 : _a.split("&").forEach((p) => {
                    params[p.split("=")[0]] = decodeURIComponent(p.split("=")[1]);
                });
                const skipPasswordCheck = toolkitGlobal_1.default.getInstance().debugBridge.disableDebugBridgePassword;
                let allowed = false;
                if (skipPasswordCheck) {
                    allowed = true;
                }
                else {
                    // If password is not set, ask permission for commands without password.
                    if (typeof params.password === "undefined" &&
                        typeof this.password === "undefined") {
                        allowed = window.confirm(`External App ${params.app} wants to execute command without password.\nCommand:\n${(params.run ||
                            params.file ||
                            "").slice(0, 100)}\nIf you do not know what it is, please click Cancel to deny.`);
                    }
                    else {
                        allowed = this.password === params.password;
                    }
                }
                if (allowed) {
                    if (params.run) {
                        try {
                            const AsyncFunction = Object.getPrototypeOf(async function () { }).constructor;
                            const f = new AsyncFunction("Zotero,window", params.run);
                            await f(Zotero, window);
                        }
                        catch (e) {
                            Zotero.debug(e);
                            window.console.log(e);
                        }
                    }
                    if (params.file) {
                        try {
                            Services.scriptloader.loadSubScript(params.file, {
                                Zotero,
                                window,
                            });
                        }
                        catch (e) {
                            Zotero.debug(e);
                            window.console.log(e);
                        }
                    }
                }
            },
            newChannel: function (uri) {
                this.doAction(uri);
            },
        };
        // @ts-ignore
        Services.io.getProtocolHandler("zotero").wrappedJSObject._extensions["zotero://ztoolkit-debug"] = debugBridgeExtension;
    }
}
exports.DebugBridge = DebugBridge;
DebugBridge.version = 2;
DebugBridge.passwordPref = "extensions.zotero.debug-bridge.password";
