// zotero-plugin-toolkit/dist/_doc.js
"use strict";
/**
 * This file is for api-extractor. Things here will appear in the docs.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuideHelper = exports.LargePrefHelper = exports.DialogHelper = exports.VirtualizedTableHelper = exports.ProgressWindowHelper = exports.FilePickerHelper = exports.ClipboardHelper = exports.ShortcutManager = exports.PatchHelper = exports.PreferencePaneManager = exports.MenuManager = exports.ReaderInstanceManager = exports.ReaderTabPanelManager = exports.LibraryTabPanelManager = exports.KeyModifier = exports.KeyboardManager = exports.ItemBoxManager = exports.ItemTreeManager = exports.FieldHookManager = exports.ExtraFieldTool = exports.ReaderTool = exports.UITool = exports.ZoteroToolkit = exports.BasicTool = void 0;
const basic_1 = require("./basic");
Object.defineProperty(exports, "BasicTool", { enumerable: true, get: function () { return basic_1.BasicTool; } });
const index_1 = require("./index");
Object.defineProperty(exports, "ZoteroToolkit", { enumerable: true, get: function () { return index_1.ZoteroToolkit; } });
const ui_1 = require("./tools/ui");
Object.defineProperty(exports, "UITool", { enumerable: true, get: function () { return ui_1.UITool; } });
const reader_1 = require("./tools/reader");
Object.defineProperty(exports, "ReaderTool", { enumerable: true, get: function () { return reader_1.ReaderTool; } });
const extraField_1 = require("./tools/extraField");
Object.defineProperty(exports, "ExtraFieldTool", { enumerable: true, get: function () { return extraField_1.ExtraFieldTool; } });
const fieldHook_1 = require("./managers/fieldHook");
Object.defineProperty(exports, "FieldHookManager", { enumerable: true, get: function () { return fieldHook_1.FieldHookManager; } });
const itemTree_1 = require("./managers/itemTree");
Object.defineProperty(exports, "ItemTreeManager", { enumerable: true, get: function () { return itemTree_1.ItemTreeManager; } });
const itemBox_1 = require("./managers/itemBox");
Object.defineProperty(exports, "ItemBoxManager", { enumerable: true, get: function () { return itemBox_1.ItemBoxManager; } });
const keyboard_1 = require("./managers/keyboard");
Object.defineProperty(exports, "KeyboardManager", { enumerable: true, get: function () { return keyboard_1.KeyboardManager; } });
Object.defineProperty(exports, "KeyModifier", { enumerable: true, get: function () { return keyboard_1.KeyModifier; } });
const libraryTabPanel_1 = require("./managers/libraryTabPanel");
Object.defineProperty(exports, "LibraryTabPanelManager", { enumerable: true, get: function () { return libraryTabPanel_1.LibraryTabPanelManager; } });
const readerTabPanel_1 = require("./managers/readerTabPanel");
Object.defineProperty(exports, "ReaderTabPanelManager", { enumerable: true, get: function () { return readerTabPanel_1.ReaderTabPanelManager; } });
const readerInstance_1 = require("./managers/readerInstance");
Object.defineProperty(exports, "ReaderInstanceManager", { enumerable: true, get: function () { return readerInstance_1.ReaderInstanceManager; } });
const menu_1 = require("./managers/menu");
Object.defineProperty(exports, "MenuManager", { enumerable: true, get: function () { return menu_1.MenuManager; } });
const preferencePane_1 = require("./managers/preferencePane");
Object.defineProperty(exports, "PreferencePaneManager", { enumerable: true, get: function () { return preferencePane_1.PreferencePaneManager; } });
const shortcut_1 = require("./managers/shortcut");
Object.defineProperty(exports, "ShortcutManager", { enumerable: true, get: function () { return shortcut_1.ShortcutManager; } });
const clipboard_1 = require("./helpers/clipboard");
Object.defineProperty(exports, "ClipboardHelper", { enumerable: true, get: function () { return clipboard_1.ClipboardHelper; } });
const filePicker_1 = require("./helpers/filePicker");
Object.defineProperty(exports, "FilePickerHelper", { enumerable: true, get: function () { return filePicker_1.FilePickerHelper; } });
const progressWindow_1 = require("./helpers/progressWindow");
Object.defineProperty(exports, "ProgressWindowHelper", { enumerable: true, get: function () { return progressWindow_1.ProgressWindowHelper; } });
const virtualizedTable_1 = require("./helpers/virtualizedTable");
Object.defineProperty(exports, "VirtualizedTableHelper", { enumerable: true, get: function () { return virtualizedTable_1.VirtualizedTableHelper; } });
const dialog_1 = require("./helpers/dialog");
Object.defineProperty(exports, "DialogHelper", { enumerable: true, get: function () { return dialog_1.DialogHelper; } });
const largePref_1 = require("./helpers/largePref");
Object.defineProperty(exports, "LargePrefHelper", { enumerable: true, get: function () { return largePref_1.LargePrefHelper; } });
const patch_1 = require("./helpers/patch");
Object.defineProperty(exports, "PatchHelper", { enumerable: true, get: function () { return patch_1.PatchHelper; } });
const guide_1 = require("./helpers/guide");
Object.defineProperty(exports, "GuideHelper", { enumerable: true, get: function () { return guide_1.GuideHelper; } });
